### PR TITLE
Ripgrep search

### PR DESCRIPTION
**Pull request by Uzay-G**
Partly fixes https://github.com/archivy/archivy/issues/13.

Adds a [ripgrep](https://github.com/BurntSushi/ripgrep) backend for search, which is much more lightweight than ES, albeit much more limited in its functionality.

### CODE REVIEW

**Comment by Uzay-G** [comment]
@srevinsaju could you review this when you have time? I probably have to polish some stuff :)

**Comment by srevinsaju** [comment]
Looks awesome, will test it out soon!

**Comment by srevinsaju** [comment]
I wanted to ask.. is it possible to keep Elastic Search along with ripgrep? Its fine if its extra hardwork, but if we could get both of them, it would give the user some choice + adding new backends would be easier in future. What is your opinion @Uzay-G ?

**Comment by Uzay-G** [comment]
Small improvement / fix I'm not sure how to implement for this PR:

Currently the ripgrep search returns a `title` attribute based on the filename of the file the match was found in, because this is much faster than reading from the file for its title attribute. However, this returns a slightly sanitized name, which isn't perfect, and it means the titles returned by ES and ripgrep are not exactly the same.

**Comment by Uzay-G** [comment]
> ~I wanted to ask.. is it possible to keep Elastic Search along with ripgrep? If its extra hardwork, we can chuck it,.. but if we could get both of them, it would give the user some choice + adding new backends would be easier in future. What is your opinion @Uzay-G ?~, ah nvm, you are smart!

:joy: Yeah I kept ES because it is quite a nice backend and offers a bit more functionality.

I've also tried my best to make this PR fully backwards compatible despite the addition of the new `engine` configuration by guessing the default `engine` on runtime.

File: archivy/cli.py [hunk 1]
```diff
@@ -47,15 +47,21 @@ def init(ctx):
         default=str(Path(".").resolve()),
     )
 
-    es_enabled = click.confirm(
-        "Would you like to enable Elasticsearch? For this to work "
-        "when you run archivy, you must have ES installed."
-        "See https://archivy.github.io/setup-search/ for more info."
+    desires_search = click.confirm(
+        "Would you like to have search on your knowledge base contents?"
```
Comment (srevinsaju) [inline archivy/cli.py:51]: Can we better phrase this? It looks like something is missing in this line?

Perhaps 'indexing' is the right word, or a ',' is missing?

**Review by srevinsaju — APPROVED** [review]

**Comment by srevinsaju** [comment]
I did a code review, looks good. I will test this branch in a while.

**Comment by srevinsaju** [comment]
> > ~I wanted to ask.. is it possible to keep Elastic Search along with ripgrep? If its extra hardwork, we can chuck it,.. but if we could get both of them, it would give the user some choice + adding new backends would be easier in future. What is your opinion @Uzay-G ?~, ah nvm, you are smart!
> 
>  Yeah I kept ES because it is quite a nice backend and offers a bit more functionality.
> 
> I've also tried my best to make this PR fully backwards compatible despite the addition of the new `engine` configuration by guessing the default `engine` on runtime.

Yea, thats awesome.

File: archivy/cli.py (hunk 1 above)
Comment (srevinsaju) [reply archivy/cli.py:51]: * Would you like to have indexing on your know....
* "Would you like to have 'search' on your knowledge base contents?"
* etc etc, does it sound right? I am not sure either...

Comment (Uzay-G) [reply archivy/cli.py:51]: Hmm yeah you're right, it could be improved. What would you propose?

Comment (Uzay-G) [reply archivy/cli.py:51]: - "Would you like to enable search on your knowledge base contents?"
Maybe that could work?

Comment (srevinsaju) [reply archivy/cli.py:51]: > * "Would you like to enable search on your knowledge base contents?"

Yea, this looks perfect.

**Comment by srevinsaju** [comment]
Tested, works awesome. Thanks a million!!

**Comment by srevinsaju** [comment]
Btw, if ever interests you, you would like to see about https://github.com/phiresky/ripgrep-all which includes indexing PDFs too, :smile: , in future if you would like to add a PDF archiving feature too :smile:

**Pull request merged**